### ISSUE TITLE
Performance improvement in SPI

### ISSUE THREAD

**Issue by tgiacchi** [carlk3/FreeRTOS-FAT-CLI-for-RPi-Pico#27]
Chaning the spi_set_format call in my_spi_init to use SPI_CPHA_1 seems to bring about a 10% improvement in performance.

**Comment by carlk3** [comment]
I'm surprised that works at all. ChaN [says](http://elm-chan.org/docs/mmc/mmc_e.html#spimode) 

> SPI mode 0 (CPHA=0, CPOL=0) is the proper setting to control MMC/SDC, but mode 3 (CPHA=1, CPOL=1) also works as well in most case. 

but you're using mode 1? He doesn't cite his source, unfortunately. I looked at **SD Specifications**, Part 1, _Physical Layer Specification_, Version 3.01, February 18, 2010, and I couldn't find any mention of the SPI mode, and the bus timing diagrams are omitted in the simplified (free) versions of the SD Physical Layer specification. The Simplified Specifications are a subset of the complete SD Specifications, which I don't have. If you have access to the full Specifications, you might want to check whether SPI mode 1 operation is compliant.

**Comment by tgiacchi** [comment]
I made the change based on looking at the rp2040 datasheet and scope views. I originally did it for the noOs version of the code to get that little extra performance I needed.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by tgiacchi** [comment]
Quick question.  Have you gotten this to work with pico-w doing tcp? If you have an example that would be great. I've been pulling my hair out for days.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by carlk3** [comment]
> I made the change based on looking at the rp2040 datasheet and scope views. I originally did it for the noOs version of the code to get that little extra performance I needed. 

How are you measuring the performance improvement? I can't imagine how changing the phase could significantly change the throughput unless you are also increasing the baud rate.

**Comment by tgiacchi** [comment]
No baud rate increase. Writing data to the file and get usec time at start and end. Pretty simple. I am running at 25mhz exactly.  Tested on 5 different cards from several different manufacturers.

When I find some free time I'll hook up the scope again and do a capture and send it along.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by carlk3** [comment]
> No baud rate increase. Writing data to the file and get usec time at start and end. Pretty simple. I am running at 25mhz exactly. Tested on 5 different cards from several different manufacturers. When I find some free time I'll hook up the scope again and do a capture and send it along. 

Do you have a theory on how changing the clock phase could improve the data transfer rate?

There are many variables that could affect the timing. (See [Performance Tuning Tips](https://github.com/carlk3/no-OS-FatFS-SD-SDIO-SPI-RPi-Pico#appendix-d-performance-tuning-tips). For example, the file might span a segment on the SD card. Are you formatting the cards between tests?

**Comment by tgiacchi** [comment]
The ph0 drops the clock for a longer period.  Not sure why. The same sd card, same file that is preallocated before hand. So the same sectors are being read and written.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by tgiacchi** [comment]
Yes, clock is low for a longer period. As far a file size, 8 to 16mb.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by carlk3** [comment]
> Yes, clock is low for a longer period.

How does having the clock is low for a longer period improve the performance? Is the overall clock cycle period shorter? I can't understand how it can be moving data faster unless the frequency goes up or somehow there's less host wait time (normally due to card busy time).

**Comment by tgiacchi** [comment]
Let me rephrase.  In ph0 the clock is low longer making each xfer take longer. Ph1 doesn't have that ossue

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by carlk3** [comment]
Ah, so the frequency does change. What if you raise the `baud_rate` in ph0 instead of switching to ph1?

**Comment by carlk3** [comment]
I don't understand how the frequency can change. From _RP2040 Datasheet_:

> 4.4.3.6.1. Bit rate generation
The serial bit rate is derived by dividing down the input clock, SSPCLK. The clock is first divided by an even prescale
value CPSDVSR in the range 2-254, and is programmed in SSPCPSR. The clock is divided again by a value in the range 1-256, that is 1 + SCR, where SCR is the value programmed in SSPCR0.

where SSPCLK is wired to `clk_peri`. By default `clk_peri` is attached directly to the system clock. So I would guess that the serial bit rate would be some constant ratio to the `clk_peri`. How would changing the clock phase affect that?

**Comment by tgiacchi** [comment]
It's not the frequency per se. It's the time the chip takes to start sending data for each packet. This weekend I'll have time to pull some traces that show what I am talking about.  A picture is worth a thousand words.

Btw, I appreciate all the help. I'm way behind on my Flux copier so getting things going with rtos really helped.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by carlk3** [comment]
I tried using 
```C
        spi_set_format(spi_p->hw_inst, 8, SPI_CPOL_0, SPI_CPHA_1, SPI_MSB_FIRST);
``` 
and I can't get the card to initialize. Even at 122,070 Hz, it gets CRC errors on CMD0, "Go Idle State", which is the first command.

**Comment by tgiacchi** [comment]
Wow. That's odd. Every card I use works.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by carlk3** [comment]
It's the [Silicon Power 3D NAND U1 32GB microSD card](https://www.amazon.com/gp/product/B07RSXSYJC/) that refuses to initialize with `SPI_CPHA_1`. So, I switched to a [SanDisk Ultra 16GB](https://www.westerndigital.com/products/memory-cards/sandisk-ultra-uhs-i-microsd?sku=SDSQUNC-016G-AN6MA). That does work, so I ran some benchmarks with [big_file_test.c](https://github.com/carlk3/FreeRTOS-FAT-CLI-for-RPi-Pico/blob/master/examples/command_line/tests/big_file_test.c) and [bench.c](https://github.com/carlk3/FreeRTOS-FAT-CLI-for-RPi-Pico/blob/master/examples/command_line/tests/bench.c). My results are summarized here: [SPI_CPHA test results](https://docs.google.com/spreadsheets/d/1lDVyZO5_2BJLS-WF96ADBPmk8W7aX-u8vp1t2ejg48g/edit?usp=sharing).

In general, with `SPI_CPHA_1` I found:

[big_file_test.c](https://github.com/carlk3/FreeRTOS-FAT-CLI-for-RPi-Pico/blob/master/examples/command_line/tests/big_file_test.c), 20 MiB file:

- On average (three trials), it took 6.40% more time to write the file
- No change in time to read a 20 MiB file

[bench.c](https://github.com/carlk3/FreeRTOS-FAT-CLI-for-RPi-Pico/blob/master/examples/command_line/tests/bench.c), 5 MiB file:

- Write speed decreased 5.50% for a 5 MiB file
- Average write latency increased 5.35%
- Read speed decreased 1.18% for a 5 MiB file
- Average read latency increased 1.09%

Writing to SD cards is far from deterministic, there is a lot variance in these results, and I only tested with the [SanDisk Ultra 16GB](https://www.westerndigital.com/products/memory-cards/sandisk-ultra-uhs-i-microsd?sku=SDSQUNC-016G-AN6MA), so more testing would be needed to really quantify the differences, but I'm not seeing the performance  improvement.

**Comment by tgiacchi** [comment]
So strange. Guess it depends on the card.

On other issues, still getting hangs and unreliable behavior.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by tgiacchi** [comment]
Probably not worth worrying about then.

I've just about given up on getting wifi and sd card stuff working together.  Absolutely unstable.

If you have someplace I can post my code , maybe I'm missing something

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by carlk3** [comment]
> If you have someplace I can post my code , maybe I'm missing something

You can always [fork ](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/fork-a-repo) this repository on GitHub.

**Comment by tgiacchi** [comment]
I created a repo and gave you access.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
